src/rospack.cpp

```diff
@@ -1994,23 +1994,33 @@ Rosstackage::writeCache()
   }
   else
   {
-    char tmp_cache_dir[PATH_MAX];
-    char tmp_cache_path[PATH_MAX];
-    strncpy(tmp_cache_dir, cache_path.c_str(), sizeof(tmp_cache_dir));
+    int len = cache_path.size() + 1;
+    char *tmp_cache_dir = (char *)malloc(len);
+    strncpy(tmp_cache_dir, cache_path.c_str(), len);
+    // make sure tmp_cache_dir is NULL terminated
+    tmp_cache_dir[len - 1] = '\0';
 #if defined(_MSC_VER)
     // No dirname on Windows; use _splitpath_s instead
+    char tmp_cache_path[PATH_MAX];
     char drive[_MAX_DRIVE], dir[_MAX_DIR], fname[_MAX_FNAME], ext[_MAX_EXT];
     _splitpath_s(tmp_cache_dir, drive, _MAX_DRIVE, dir, _MAX_DIR, fname, _MAX_FNAME,
                  ext, _MAX_EXT);
     char full_dir[_MAX_DRIVE + _MAX_DIR];
     _makepath_s(full_dir, _MAX_DRIVE + _MAX_DIR, drive, dir, NULL, NULL);
     snprintf(tmp_cache_path, sizeof(tmp_cache_path), "%s\\.rospack_cache.XXXXXX", full_dir);
 #elif defined(__MINGW32__)
+    char tmp_cache_path[PATH_MAX];
     char* temp_name = tempnam(dirname(tmp_cache_dir),".rospack_cache.");
     snprintf(tmp_cache_path, sizeof(tmp_cache_path), temp_name);
     delete temp_name;
 #else
-    snprintf(tmp_cache_path, sizeof(tmp_cache_path), "%s/.rospack_cache.XXXXXX", dirname(tmp_cache_dir));
+    char *tmp_cache_path = NULL;
+    char *temp_dirname = strdup(dirname(tmp_cache_dir));
```

Comment:
What is the purpose of `temp_dirname`? Why not use `tmp_cache_dir` which it is copied from?